### ISSUE TITLE
not all msg published

### ISSUE THREAD

**Issue by sangshuduo** [dist-sys/mqttloader#10]
Hello,

I used mqttloader to test hivemq-ce, now I see only 9980 msg be published and only 3368 msg be received. I think at least 10000 msg should be published if not all msg received. Am I right?

Thanks,

$ ./mqttloader -b tcp://127.0.0.1:1883 -p 1000 -s 2 -m 10 -t "t/a" -pq 2 -sq 2

\-----Publisher-----

Maximum throughput\[msg/s\]: 2315

Average throughput\[msg/s\]: 1425.7142857142858

Number of published messages: 9980

Throughput\[msg/s\]: 1403, 2315, 1225, 1900, 1248, 1848, 41

\-----Subscriber-----

Maximum throughput\[msg/s\]: 800

Average throughput\[msg/s\]: 374.22222222222223

Number of received messages: 3368

Throughput\[msg/s\]: 113, 137, 317, 233, 358, 257, 753, 800, 400

Maximum latency\[ms\]: 6126

Average latency\[ms\]: 3943.5136579572445

**Comment by nabon** [comment]
Thank you for your feedback!
Since using QoS 2 could cause congestion on the broker, I guess part of messages were dropped due to the full of the internal queues, in regards to the received messages.

But so far, I could not reproduce the loss of published messages.
Would you tell me the HiveMQ version, and also the configuration, if you changed it from the default values?

**Comment by sangshuduo** [comment]
I'm using HiveMQ community edition 2020.4 docker image. I didn't change any default value. I run it on a VirtualBox VM with 8GB RAM and OS is Ubuntu 18.04. I think it should be easy to reproduce.

I run it again without quality specified. The result is following:

$ ./mqttloader -b tcp://127.0.0.1:1883 -p 1000 -s 2 -m 10 -t "t/a"

\-----Publisher-----

Maximum throughput\[msg/s\]: 4251

Average throughput\[msg/s\]: 2492.5

Number of published messages: 9970

Throughput\[msg/s\]: 335, 3336, 4251, 2048

\-----Subscriber-----

Maximum throughput\[msg/s\]: 8747

Average throughput\[msg/s\]: 4985.0

Number of received messages: 19940

Throughput\[msg/s\]: 750, 6351, 8747, 4092

Maximum latency\[ms\]: 624

Average latency\[ms\]: 40.44428284854564

**Comment by nabon** [comment]
Thank you for the information! I could reproduce the issue.

I found that the part accessing a `ScheduledFuture` class's instance occasionally gets stuck.
I'm still investigating what occurred inside java.util.concurrent libraries, but I think I've fixed the problem itself.

The attached file below is the modified binary.
[mqttloader.zip](https://github.com/dist-sys/mqttloader/files/5151379/mqttloader.zip)

While investigating, I found some other parts to improve the codes, so I'll reflect them and soon release version 0.6.3.
